crates/wasi/src/preview2/ip_name_lookup.rs

```diff
@@ -5,35 +5,73 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 use crate::preview2::{spawn_blocking, AbortOnDropJoinHandle, SocketError, WasiView};
 use anyhow::Result;
 use std::mem;
-use std::net::{Ipv6Addr, ToSocketAddrs};
+use std::net::{IpAddr, Ipv6Addr, ToSocketAddrs};
 use std::pin::Pin;
 use std::str::FromStr;
 use std::vec;
 use wasmtime::component::Resource;
 
-use super::network::{from_ipv4_addr, from_ipv6_addr};
+use super::WasiCtx;
 
 pub enum ResolveAddressStream {
-    Waiting(AbortOnDropJoinHandle<Result<Vec<IpAddress>, SocketError>>),
-    Done(Result<vec::IntoIter<IpAddress>, SocketError>),
+    Waiting(AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>),
+    Done(std::io::Result<vec::IntoIter<IpAddr>>),
 }
 
-#[async_trait::async_trait]
-impl<T: WasiView> Host for T {
+/// A trait for providing a `IpNameLookup` implementation.
+pub trait WasiNetworkView: Send {
+    /// Given a name, resolve to a list of IP addresses
     fn resolve_addresses(
         &mut self,
-        network: Resource<Network>,
         name: String,
-    ) -> Result<Resource<ResolveAddressStream>, SocketError> {
-        let network = self.table().get(&network)?;
+    ) -> AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>;
+}
 
-        let host = parse(&name)?;
+/// The default implementation for `WasiIpNameLookupView`.
+#[derive(Debug, Clone, Default)]
+pub struct SystemNetwork {
+    allowed: bool,
+}
 
-        if !network.allow_ip_name_lookup {
-            return Err(ErrorCode::PermanentResolverFailure.into());
+impl SystemNetwork {
+    /// Create a new `SystemIpNameLookup`
+    pub fn new(ctx: &WasiCtx) -> Self {
+        Self {
+            allowed: ctx.allowed_network_uses.ip_name_lookup,
         }
+    }
+}
```

Comment:
Could this perhaps manifest as a `impl WasiNetworkView for WasiCtx` to avoid a second type?

Reply:
(then it might be neat to have default implementations of `fn network_view` perhaps?)

Reply:
Great idea!